tools/lisa-build-asset

```diff
@@ -57,25 +83,89 @@ def list_assets(recipe_dir):
         if script.endswith('.recipe')
     )
 
+def get_recipe_name(asset):
+    return f'{asset}.recipe'
+
 def get_asset_recipe(asset, recipe_dir):
-    return os.path.join(recipe_dir, '{}.recipe'.format(asset))
+    return os.path.join(recipe_dir, get_recipe_name(asset))
+
+def get_asset_property(asset, arch, property, **kwargs):
+    # Read asset's recipe to get a property value.
+    action = f"printf '%s\\n' ${{{property}[@]}}"
+    val = make(asset, arch, [action], **kwargs, get_output=True)
+    return val[action].splitlines()
+
+@contextmanager
+def make_chroot(asset, arch, build_dir=None, recipe_dir=None, arch_chroot_dir=None):
+    # Create a chroot at arch_chroot_dir as a contextmanager.
+    recipe = get_asset_recipe(asset, recipe_dir)
+    env = get_env(asset, arch, None, recipe, build_dir)
+
+    # List the required Alpine packages.
+    alpine_dependencies = get_asset_property(
+                                asset, arch, "ALPINE_BUILD_DEPENDENCIES",
+                                build_dir=build_dir,
+                                recipe_dir=recipe_dir,
+                                arch_chroot_dir=arch_chroot_dir
+                            )
+    alpine_dependencies_args = ('-p', ' '.join(map(quote, alpine_dependencies))) if alpine_dependencies else None
+
+    # List the required env variables.
+    used_env_variables = get_used_env_var(get_asset_recipe(asset, recipe_dir), env)
+    used_env_variables_args = ('-k', ' '.join(used_env_variables)) if used_env_variables else None
+
+    # Install the chroot.
+    subprocess.check_call(['sudo', os.path.join(get_alpine_dir(), 'alpine-chroot-install'),
+            '-d', arch_chroot_dir,
+            '-i', build_dir,
+            '-a', QEMU_ARCH_NAMES.get(arch, arch),
+            *alpine_dependencies_args,
+            *used_env_variables_args],
+            cwd=build_dir)
+
+    # Manually bind additionnal the recipe directory to the chroot.
+    subprocess.check_call(['sudo', 'mkdir', '-p', arch_chroot_dir + recipe_dir])
+    subprocess.check_call(['sudo', 'mount', '--bind', recipe_dir, arch_chroot_dir + recipe_dir])
+
+    try:
+        yield arch_chroot_dir
+    finally:
+        # Clean-up the chroot
+        subprocess.check_call([os.path.join(arch_chroot_dir, 'destroy'), '--remove'])
+
+def make(asset, arch, actions, toolchain=None, build_dir=None, recipe_dir=None, arch_chroot_dir=None, get_output=False):
+    if arch_chroot_dir:
+        # Since the build is native, no need for a toolchain.
+        toolchain = None
+
+    if get_output:
+        # make() is sometimes expected to return the output of the 'actions'.
+        # If so, use check_output() to exectute the 'actions'.
+        output = dict()
+        def run_cmd(*args, **kwargs):
+            output[action] = subprocess.check_output(*args, **kwargs).decode('utf-8')
```

Comment:
It took me some time to understand, but yes, I didn't realize how dodgy this was

Reply:
Somewhat related, there is a Python gotcha related to that:
```
functions = [
    lambda: print(i)
    for i in range(2)
]

for f in functions:
    f()
```
This will print:
```
1
1
```

This is because `i` is referenced by name in the function, and the name will get resolved at runtime by looking it up in enclosing scopes until it's found. When the loop finishes, the `i` name is bound to the value `1`, and any closure defined in this scope will therefore get `1` when trying to refer to `i`.

A workaround is to use a function factory like that:
```def make_f(i):
    return lambda: print(i)

functions = [
    make_f(i)
    for i in range(2)
]

for f in functions:
    f()
```

Or if you are able to use `def`, you can use a parameter with default value, which will introduce a new name `i` in the scope of `f`, bound to the value that the loop's `i` had when the function was defined:
```
functions = []
for i in range(2):
    def f(i=i):
        print(i)
    functions.append(f)

for f in functions:
    f()
```